### PR TITLE
complete data layout transform

### PR DESCRIPTION
**Pull request by jacquesqiao**
fix: https://github.com/PaddlePaddle/Paddle/issues/7436

### CODE REVIEW

File: paddle/framework/op_kernel_type.h [hunk 1]
```diff
@@ -85,9 +85,14 @@ inline std::string KernelTypeToString(const OpKernelType& kernel_key) {
...
 }
 
+inline bool NeedTransformLayout(const DataLayout& l, const DataLayout& r) {
+  return l != DataLayout::kAnyLayout && r != DataLayout::kAnyLayout && l != r;
+}
+
 inline bool TransFromNeeded(const OpKernelType& l, const OpKernelType& r) {
   return (!platform::places_are_same_class(l.place_, r.place_)) ||
-         (l.data_type_ != r.data_type_) || (l.data_layout_ != r.data_layout_);
+         (l.data_type_ != r.data_type_) ||
```
Comment (dzhwinter) [inline paddle/framework/op_kernel_type.h]: **I believe we are in the wrong way.** 

let's start from a simple convolution case:
data -> conv2d -> batch_norm -> fc -> softmax
we may have a quite different layout pipelines choices.
NHWC->NCHW -> NCHW -> NCHW
NHWC->NHWC -> NCHW -> NCHW
NHWC->NHWC -> NHWC -> NCHW
....
NCHW->NCHW -> NCHW -> NCHW
NCHW->NHWC -> NCHW -> NCHW
....
All of them are legal, but only one of them is the best.  in cudnn, NHWC is 20% faster than other formats.
So, the answer to select one from above, it's would be NHWC pipeline, no matter what the data format is, just transform the data into NHWC, then go through the data pipeline.

In the case above, I want to claim that **we do not run on mixed devices, we can run on different device type.**.  
It means the **data pipeline layout** can be determined once we have the device information, we only need to transform data before or after the data pipeline. 
data pipeline refer to `conv2d -> batch_norm -> fc -> softmax`, without data input and output.

Comment (dzhwinter) [reply paddle/framework/op_kernel_type.h]: To be clear, we support multiple device, but not mixed device.

File: paddle/framework/data_transform.cc [hunk 2]
```diff
@@ -15,18 +15,58 @@ limitations under the License. */
...
                    const Tensor& input_tensor, Tensor* out) {
+  // in_ptr is used to store the input of data transform function.
+  // it need to be deleted by free_tmp_tensor because ip_ptr may be
+  // equal to &input_tensor and input_tensor should not be deleted here.
+  const Tensor* in_ptr = &input_tensor;
+
+  // out_ptr is used to store the result of data transform function.
+  // when one transform is done, the in_ptr should be deleted and the out_ptr
+  // should be assigned to in_ptr for next data transformation.
+  Tensor* out_ptr = new Tensor();
```
Comment (dzhwinter) [inline paddle/framework/data_transform.cc]: new and delete inside DataTransform looks like awful.  Furthermore, the memory optimizer cannot track these actions. 
How about create 3 Tensor(for each channel transform) in scope?

File: paddle/framework/data_transform.cc [hunk 3]
```diff
@@ -15,18 +15,58 @@ limitations under the License. */
...
+
+  // do layout transform
+  if (NeedTransformLayout(expected_kernel_type.data_layout_,
+                          kernel_type_for_var.data_layout_)) {
+    TransDataLayout(kernel_type_for_var, expected_kernel_type, *in_ptr,
+                    out_ptr);
+    free_tmp_tensor(&input_tensor, in_ptr);
```
Comment (dzhwinter) [inline paddle/framework/data_transform.cc]: free_tmp_tensor looks awful too...

File: paddle/framework/op_kernel_type.h (hunk 1 above)
Comment (jacquesqiao) [reply paddle/framework/op_kernel_type.h]: yes, I Agree with you, we have a data process chain, and there should be some best practice. On the other hand, I also think that the current data transform just does transform according to expected_kernel_type and actual_kernel_type_for_var, it does not consider what kind of expected_kernel_type it should use, I think that should be implemented in GetExpectedKernelType by the user or by framework. So it does not conflict with the current implementation.

File: paddle/framework/data_transform.cc (hunk 2 above)
Comment (jacquesqiao) [reply paddle/framework/data_transform.cc]: Ok, let me think a better way to do this~

File: paddle/framework/data_transform.cc (hunk 3 above)
Comment (jacquesqiao) [reply paddle/framework/data_transform.cc]: updated

File: paddle/framework/data_transform.cc [hunk 4]
```diff
@@ -15,18 +15,43 @@ limitations under the License. */
...
   if (!platform::is_same_place(kernel_type_for_var.place_,
                                expected_kernel_type.place_)) {
-    DeviceTransform(input_tensor, expected_kernel_type.place_, out);
+    DeviceTransform(in, expected_kernel_type.place_, &out);
+    transformed = true;
+    PassTensorData(&out, &in);
   }
-  PADDLE_ENFORCE_NOT_NULL(out, "out should not be null");
+
+  PADDLE_ENFORCE(transformed, "no transform is done, please check!");
```
Comment (dzhwinter) [inline paddle/framework/data_transform.cc]: Comment follow UpperCase, and in written English.
"No transform is applied"

File: paddle/framework/data_transform.cc [hunk 5]
```diff
@@ -15,18 +15,43 @@ limitations under the License. */
...
+
 void DataTransform(const OpKernelType& expected_kernel_type,
                    const OpKernelType& kernel_type_for_var,
-                   const Tensor& input_tensor, Tensor* out) {
+                   const Tensor& input_tensor, Tensor* output_tensor) {
+  bool transformed = false;
+  Tensor in;
+  in.ShareDataWith(input_tensor);
+  Tensor out;
+
```
Comment (dzhwinter) [inline paddle/framework/data_transform.cc]: Where is the datatype transform ?

File: paddle/framework/data_transform.cc [hunk 6]
```diff
@@ -15,18 +15,43 @@ limitations under the License. */
 #include "paddle/framework/data_transform.h"
 
 #include "paddle/framework/data_device_transform.h"
+#include "paddle/framework/data_layout_transform.h"
 
 namespace paddle {
 namespace framework {
 
+static void PassTensorData(Tensor* from, Tensor* to) {
```
Comment (dzhwinter) [inline paddle/framework/data_transform.cc]: remove static kerword

File: paddle/framework/data_transform.cc (hunk 5 above)
Comment (jacquesqiao) [reply paddle/framework/data_transform.cc]: in next pr
